### PR TITLE
Safetensor support

### LINKED ISSUE #182: Safetensor support
Currently only numpy file format is supported, was wondering if safetensor support was of interest.

### PR DESCRIPTION
**Pull request by dc-dc-dc**
## Proposed changes

Adds support for `safetensors` as specified [here](https://huggingface.co/docs/safetensors/index) and closes #182.

Added a bare minimum json parser for safetensor metadata parsing, still have some code to cleanup / simplify. But wanted to make a draft PR to get some initial comments on code placement / file structure and other nuances as well as any changes that maintainers might seek

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

**Comment by awni** [comment]
Wow very cool! Will take a look soon!

**Comment by JulianGerhard21** [comment]
Hi @dc-dc-dc - awesome, thanks a lot. Highly anticipated feature. Can it be used yet? I'd provide some feedback if so.

**Comment by awni** [comment]
I'm wondering about the decision to write a custom minimal json parser vs use something open source. For example https://github.com/nlohmann/json is header only and MIT License so we could include it the same way we include Doctest. (fetch it in CMake). Any thoughts on the pros/cons?

**Comment by dc-dc-dc** [comment]
I saw some people recommending this lib when I was originally looking into adding json support, originally felt that minimum support for strings / numbers would be all thats needed but looks like `__metadata__` will require more data type support.

I will mess with utilizing this lib and see how it fits

**Comment by dc-dc-dc** [comment]
Update: decided to remove the custom json parser and utilize `nlohmann/json`. I am having some issues trying to get cmake FetchContent to link the lib so for now am using find_library after installing using homebrew. If someone has some more experience with cmake build and knows how to resolve this I would appreciate it 😄

**Comment by awni** [comment]
For an API what do you think about the following:

A single load function which takes a file name.
```
mx.load(filename, format=None)
```

If `format is None` (default) it attempts to infer the file type based on the extension (.npy, .npz, .safetensor, etc) which will nicely for 99% of cases. But if the format specifier is provided (for example ``format="np"`) then it overrides the file type inference. CC @jagrit06 @angeloskath in case they have thoughts on that.

**Comment by awni** [comment]
> I am having some issues trying to get cmake FetchContent to link the lib so for now am using find_library after installing using homebrew.

I added it with Cmake and pushed to your branch. I might try and find a way to only get the header since the whole repo is a bit cumbersom, but for now that should work.

**Comment by julien-c** [comment]
would be awesome to have a `save_safetensors` indeed!

**Comment by dc-dc-dc** [comment]
Updated load to take a format else infer from the file extension

Added `npy`, `npz`, and `safetensors` as options

```python
import mlx.core as mx
# load safetensors format from different extension
mx.load("./temp.safe", "safetensors")
```

File: mlx/safetensor.cpp
Comment (angeloskath) [inline mlx/safetensor.cpp:145]: Out of curiosity, why not pass the `retain_graph` argument like in `save`?

Comment (dc-dc-dc) [reply mlx/safetensor.cpp:145]: Thanks for pointing this out, initially just set it to false as I was working through the code. Going back now to add this argument to the function call.

**Comment by awni** [comment]
@dc-dc-dc what's the state of this PR? Ready for review again! I'm really stoked to get this landed 😄

**Comment by awni** [comment]
To the extent that adding `_metadata_` won't break the API then I'm fine adding it later if/when requested.

For `save_safetensors` it should be ok, since it would be an optional kwarg anywa.

For `load` I think it should also be fine if the idea is to put it in the dictionary return value with the key `_meta_`.

If that sounds right, then it makes sense to me to add it in later. The C++ API would probably break but that's much more rarely used.

**Comment by dc-dc-dc** [comment]
Yeah, it wouldn’t be a breaking change from python if added only c++, an easy addition down the line too. 

With that in mind then this is ready for review 👀

**Comment by dc-dc-dc** [comment]
@awni ready

File: python/src/ops.cpp
Comment (awni) [inline python/src/ops.cpp:2954]: For consistency with `savez` should we do this as `args` (with default names) and `kwargs` - any reason not to follow that convention?

File: python/src/ops.cpp
Comment (awni) [inline python/src/ops.cpp:2956]: Can you follow the same default as `save` for this? We should in general not retain the graph unless it's a save in a function transform.

See e.g. https://github.com/ml-explore/mlx/pull/215/files#diff-8a607e80673e9f8e9c1e73a5aba9d6f89ac56ad34765e4022d5130f890776cacR2870-R2873

File: python/src/ops.cpp
Comment (awni) [inline python/src/ops.cpp:2961]: Please include a link about the format in the docstring `https://huggingface.co/docs/safetensors/index`  since it's still a pretty new format.

File: python/tests/test_load.py
Comment (awni) [inline python/tests/test_load.py:67]: The C++ and Python tests are nicely done 👍

File: python/src/ops.cpp
Comment (awni) [inline python/src/ops.cpp:2941]: The docstring needs to be updated in a couple places to mention `.safetensor` format as well.

Also for consistent use of double backticks around npy, npz, etc

File: python/src/load.cpp
Comment (awni) [inline python/src/load.cpp:443]: 😅  I had no idea we supported `pathlib.Path`, that's awesome 👍

Comment (awni) [reply python/src/load.cpp:443]: Probably good to add a test which also loads the temp file name in a `pathlib.Path` to have that case covered for the future.

File: mlx/mlx.h
Comment (awni) [inline mlx/mlx.h:11]: I don't think we want that header exposed by default in the public API. If someone wants those details they can include `mlx/safetensor.h` themselves.

**Review by awni — CHANGES_REQUESTED** [review]
This is really nicely done!!

I left a few comments mostly superficial stuff + a couple changes / questions on the API.

Also I think it would be good to make a subdir in `mlx` maybe called `mlx/io/` and put the `load.*` and `safetensor.*` stuff there. The top-level directory is getting a little crowded. Wdyt?

**Comment by dc-dc-dc** [comment]
> Also I think it would be good to make a subdir in `mlx` maybe called `mlx/io/` and put the `load.*` and `safetensor.*` stuff there. The top-level directory is getting a little crowded. Wdyt?

I like this, as in the future if any other file formats like gguf / pt support are desired they can be grouped

File: python/src/ops.cpp
Comment (dc-dc-dc) [reply python/src/ops.cpp:2954]: I think the thinking behind the two are a little different, npz can be a collection of tensors but safetensors is thought of as a dictionary of key to tensor. So passing in a dict for safetensors feels more in line

File: python/src/load.cpp
Comment (dc-dc-dc) [reply python/src/load.cpp:443]: Originally copied this error message from `mlx_load_helper` but after testing we don't actually support `Path` so will correct the error messages. Can come back and add support for it tho in a later PR

File: python/src/ops.cpp
Comment (awni) [reply python/src/ops.cpp:2954]: I'm cool with that.

**Comment by awni** [comment]
> ValueError: [load] Input must be a file-like object, or string

I am getting this failure when running the python tests. Could you tell me which version of python and pybind11 you are using?

**Comment by dc-dc-dc** [comment]
Python version 3.11
Pybind 2.11.1

**Comment by awni** [comment]
The loading from a file pointer fails with the error message I put above
```
                        with open(save_file_mlx, "rb") as f:
                            load_dict = mx.load(f)
```

**Comment by awni** [comment]
Presumably those tests pass for you?

My pybind is 2.10 and python 3.9.

pybind11-global           2.10.4           py39h48ca7d4_0  

I will try a higher version to see if that's the issue.

**Comment by dc-dc-dc** [comment]
Wait I am getting that error now after rebuilding, something I changed in the last commits must of broken it let me check

**Comment by awni** [comment]
Ok I will let you debug it then, lmk!

**Comment by dc-dc-dc** [comment]
@awni fixed on my local, was messing with adding path lib and when I was removing it out i also removed the istream_check 🤦‍♂️

File: mlx/io/safetensor.cpp
Comment (awni) [inline mlx/io/safetensor.cpp:8]: Could you make this a switch statement, and leave out the default case? The switch statement is kind of handy because if we add a new type the compiler will tell us if we missed adding it some place.

**Review by awni — COMMENTED** [review]
This looks great! I think it's pretty much ready to merge. I left one comment. 

Also, have you tried loading something a pytorch tensor that was saved using the hugging face safetensor python package? That should work right?

**Comment by awni** [comment]
I changed the name from `save_safetensor` to `save_safetensors`. I found the fact that the file extension `.safetensors` did not match the name to be confusing... they should be consistent.

Also made a few minor docs changes.

**Comment by awni** [comment]
I tested it works going from:

```
save_file(tensors, "model.safetensors")
```

And loading with `mx.load`

**Review by awni — APPROVED** [review]
I am good to go on this. I will wait to confirm my recent updates with you before merging it. 

PS this is super awesome can't wait to merge it!

**Comment by dc-dc-dc** [comment]
Looks good, one quick note about the json dependency it looks like they recommend fetching directly over using git due to size seen [here](https://json.nlohmann.me/integration/cmake/#fetchcontent). I don't have too much experience with cmake build so I defer to you on whether to keep the current git method or switch

**Comment by dc-dc-dc** [comment]
> Also, have you tried loading something a pytorch tensor that was saved using the hugging face safetensor python package? That should work right?

I was able to load phi-2 safetensors file from huggingface

**Comment by awni** [comment]
> Looks good, one quick note about the json dependency it looks like they recommend fetching directly over using git due to size seen here.

Yea worth switching that to one of the smaller options, nice fine.

File: python/src/ops.cpp
Comment (angeloskath) [inline python/src/ops.cpp:2933]: This is a very nit-picky comment in an amazing PR but I ask your understanding :-) .

I would move the format to be optionally a keyword argument ie after `py::pos_only()`. I imagine it is very readable to do `mx.load(my_python_stream, format="safetensors")`.

File: python/src/ops.cpp
Comment (angeloskath) [inline python/src/ops.cpp:2937]: Just the same as above. `/` coming before `format`.

**Review by angeloskath — CHANGES_REQUESTED** [review]
This is some fantastic work @dc-dc-dc !

I left a small nit-picky comment for the API of `mx.load`. I hope it 's not too annoying.

**Comment by awni** [comment]
Included @angeloskath suggestions along with the updated CMake (much faster download) 🚀 .

Can merge whenever @angeloskath approves.

**Comment by cchance27** [comment]
Wow this is gonna be so helpful as so much of the image landscape models are in safetensors hopefully it's merged soon

**Comment by awni** [comment]
@angeloskath this should be good to go, can you stamp? Your comments have been addressed, let's merge it unless there is anything else to add.

**Review by angeloskath — APPROVED** [review]
🚀

**Comment by awni** [comment]
PS @dc-dc-dc I forgot to mention would love to have you add your name to our [contributors](https://github.com/ml-explore/mlx/blob/main/ACKNOWLEDGMENTS.md#individual-contributors) (optional). Feel free to send a PR and I will merge it.

**Pull request merged**
